### ISSUE TITLE
False-positive no-untyped-def with `@overload`

### ISSUE THREAD

**Issue by bersbersbers** [python/mypy#13834]
**Bug Report**

I get a false-positive `no-untyped-def` on the non-`@overload`-decorated definition.

**To Reproduce**

Example from PEP 484:
> In regular modules, a series of @overload-decorated definitions must be followed by exactly one non-@overload-decorated definition (for the same function/method).

```python
"""False-positive no-untyped-def."""
from typing import overload

@overload
def utf8(value: None) -> None:
    ...

@overload
def utf8(value: bytes) -> bytes:
    ...

def utf8(value):
    print(value)
```

**Expected Behavior**

No error.

**Actual Behavior**

> bug.py:15: error: Function is missing a type annotation
Found 1 error in 1 file (checked 1 source file)

**Your Environment**

- Mypy version used: 0.982
- Mypy command-line flags: `--disallow-untyped-defs` (!)
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.10.7

**Comment by JelleZijlstra** [comment]
mypy is correct here. We need annotations on the overload implementation so that we can type check its body.

**Comment by bersbersbers** [comment]
> We need annotations on the overload implementation so that we can type check its body.

Hm - I wonder why `mypy` does not simply check the body once for each overload.

Anyway, assuming we can ignore PEP 484, I wonder how to implement this correctly. I mean, the point of the overloads is to separate disjoint sets of argument types, so it seems somewhat pointless to combine them again into a single type annotation that covers all overloads. Here's an example that does not work, but maybe I am approaching it the wrong way:

```python
from typing import Callable, overload

@overload
def pass_value_to_func(func: Callable[[float], None], value: float) -> None: ...
@overload
def pass_value_to_func(func: Callable[[int], None], value: int) -> None: ...

def pass_value_to_func(func: Callable[[float], None] | Callable[[int], None], value: float | int) -> None:
    func(value)

def print_float(value: float) -> None:
    print(value)

def print_int(value: int) -> None:
    print(value)

pass_value_to_func(print_float, 1.0)
pass_value_to_func(print_int, 1)
```

> bug.py:9: error: Argument 1 has incompatible type "Union[float, int]"; expected "int"
Found 1 error in 1 file (checked 1 source file)

**Comment by hauntsaninja** [comment]
Duplicate of https://github.com/python/mypy/issues/3360
